File: compare.py
"""
compare.py

Pose comparison engine for PoseSync.

This module compares two sets of MediaPipe pose landmarks
(reference pose + live pose) and returns:

- Match score
- Worst matching joints
- Feedback

Supports BOTH:

1. Raw MediaPipe landmarks
2. Dictionary landmarks

Author: PoseSync Team
"""

import math
from typing import List, Dict, Any

# -----------------------------
# SETTINGS
# -----------------------------

VISIBILITY_THRESHOLD = 0.5

# Higher weight = more important
JOINT_WEIGHTS = {
    0: 0.3,     # Nose

    11: 2.0,    # Left Shoulder
    12: 2.0,    # Right Shoulder

    13: 2.5,    # Left Elbow
    14: 2.5,    # Right Elbow

    15: 3.0,    # Left Wrist
    16: 3.0,    # Right Wrist

    23: 2.0,    # Left Hip
    24: 2.0,    # Right Hip

    25: 2.5,    # Left Knee
    26: 2.5,    # Right Knee

    27: 3.0,    # Left Ankle
    28: 3.0,    # Right Ankle
}

DEFAULT_WEIGHT = 1.0

# -----------------------------
# Landmark names
# -----------------------------

LANDMARK_NAMES = {
    0: "Nose",
    1: "Left Eye Inner",
    2: "Left Eye",
    3: "Left Eye Outer",
    4: "Right Eye Inner",
    5: "Right Eye",
    6: "Right Eye Outer",
    7: "Left Ear",
    8: "Right Ear",
    9: "Mouth Left",
    10: "Mouth Right",
    11: "Left Shoulder",
    12: "Right Shoulder",
    13: "Left Elbow",
    14: "Right Elbow",
    15: "Left Wrist",
    16: "Right Wrist",
    17: "Left Pinky",
    18: "Right Pinky",
    19: "Left Index",
    20: "Right Index",
    21: "Left Thumb",
    22: "Right Thumb",
    23: "Left Hip",
    24: "Right Hip",
    25: "Left Knee",
    26: "Right Knee",
    27: "Left Ankle",
    28: "Right Ankle",
    29: "Left Heel",
    30: "Right Heel",
    31: "Left Foot Index",
    32: "Right Foot Index",
}

# -----------------------------
# Helper functions
# -----------------------------

def extract_landmark_data(landmark):
    """
    Works with BOTH:

    Dictionary:
    {
        "x":...,
        "y":...,
        "z":...,
        "visibility":...
    }

    Raw MediaPipe landmark object.
    """

    if isinstance(landmark, dict):
        return (
            landmark["x"],
            landmark["y"],
            landmark.get("visibility", 1.0)
        )

    return (
        landmark.x,
        landmark.y,
        landmark.visibility
    )


def validate_landmarks(reference, live):
    """
    Ensures both landmark lists contain 33 landmarks.
    """

    if reference is None or live is None:
        raise ValueError("Landmark lists cannot be None.")

    if len(reference) != 33:
        raise ValueError(
            f"Reference pose has {len(reference)} landmarks. Expected 33."
        )

    if len(live) != 33:
        raise ValueError(
            f"Live pose has {len(live)} landmarks. Expected 33."
        )


def get_joint_weight(index):
    """
    Returns importance weight of a landmark.
    """

    return JOINT_WEIGHTS.get(index, DEFAULT_WEIGHT)

# -----------------------------
# Normalization
# -----------------------------

def normalize_landmarks(landmarks):
    """
    Normalize landmarks using the midpoint of the hips.

    This reduces the effect of the person standing
    slightly left/right/up/down in the camera.
    """

    left_hip_x, left_hip_y, _ = extract_landmark_data(landmarks[23])
    right_hip_x, right_hip_y, _ = extract_landmark_data(landmarks[24])

    center_x = (left_hip_x + right_hip_x) / 2
    center_y = (left_hip_y + right_hip_y) / 2

    normalized = []

    for landmark in landmarks:

        x, y, visibility = extract_landmark_data(landmark)

        normalized.append({
            "x": x - center_x,
            "y": y - center_y,
            "visibility": visibility
        })

    return normalized


# -----------------------------
# Distance calculation
# -----------------------------

def point_distance(x1, y1, x2, y2):
    """
    Euclidean distance between two landmarks.
    """

    return math.sqrt(
        (x2 - x1) ** 2 +
        (y2 - y1) ** 2
    )


# -----------------------------
# Compare every landmark
# -----------------------------

def compare_landmarks(reference, live):
    """
    Returns:

    [
        {
            landmark_index,
            landmark_name,
            distance,
            weight
        },
        ...
    ]
    """

    validate_landmarks(reference, live)

    reference = normalize_landmarks(reference)
    live = normalize_landmarks(live)

    comparisons = []

    for i in range(33):

        ref = reference[i]
        cur = live[i]

        if (
            ref["visibility"] < VISIBILITY_THRESHOLD
            or
            cur["visibility"] < VISIBILITY_THRESHOLD
        ):
            continue

        distance = point_distance(
            ref["x"],
            ref["y"],
            cur["x"],
            cur["y"]
        )

        comparisons.append({

            "index": i,

            "joint": LANDMARK_NAMES[i],

            "distance": distance,

            "weight": get_joint_weight(i)

        })

    return comparisons


# -----------------------------
# Overall similarity score
# -----------------------------

def calculate_match(comparisons):
    """
    Returns similarity score between 0 and 100.
    """

    if len(comparisons) == 0:
        return 0.0

    weighted_distance = 0
    total_weight = 0

    for item in comparisons:

        weighted_distance += (
            item["distance"] *
            item["weight"]
        )

        total_weight += item["weight"]

    average_distance = weighted_distance / total_weight

    MAX_EXPECTED_DISTANCE = 0.30

    score = (
        1 -
        (average_distance / MAX_EXPECTED_DISTANCE)
    ) * 100

    score = max(0, min(score, 100))

    return round(score, 2)
# -----------------------------
# Find the worst joints
# -----------------------------

def get_top_mistakes(comparisons, top_n=3):
    """
    Returns the top N joints with the highest weighted error.
    """

    ranked = sorted(
        comparisons,
        key=lambda item: item["distance"] * item["weight"],
        reverse=True
    )

    return ranked[:top_n]


# -----------------------------
# Direction helper
# -----------------------------

def get_direction(reference_joint, live_joint):
    """
    Determines how the user should move the joint.
    """

    dx = reference_joint["x"] - live_joint["x"]
    dy = reference_joint["y"] - live_joint["y"]

    THRESHOLD = 0.03

    directions = []

    # Remember:
    # Smaller y = higher on screen

    if dy < -THRESHOLD:
        directions.append("lower")

    elif dy > THRESHOLD:
        directions.append("raise")

    if dx < -THRESHOLD:
        directions.append("move right")

    elif dx > THRESHOLD:
        directions.append("move left")

    return directions


# -----------------------------
# Feedback generation
# -----------------------------

def generate_feedback(reference, live, mistakes):
    """
    Generates readable feedback.
    """

    reference = normalize_landmarks(reference)
    live = normalize_landmarks(live)

    feedback = []

    for mistake in mistakes:

        index = mistake["index"]

        joint_name = LANDMARK_NAMES[index]

        ref = reference[index]
        cur = live[index]

        directions = get_direction(ref, cur)

        if len(directions) == 0:
            message = f"Adjust your {joint_name.lower()}."

        elif len(directions) == 1:
            message = f"{directions[0].capitalize()} your {joint_name.lower()}."

        else:
            message = (
                f"{directions[0].capitalize()} and "
                f"{directions[1]} your {joint_name.lower()}."
            )

        feedback.append(message)

    return feedback


# -----------------------------
# Main function
# -----------------------------

def compare_pose(reference_landmarks, live_landmarks):
    """
    Main function to compare two poses.

    Returns:

    {
        "score": float,
        "worst_joints": [...],
        "feedback": [...]
    }
    """

    comparisons = compare_landmarks(
        reference_landmarks,
        live_landmarks
    )

    score = calculate_match(comparisons)

    mistakes = get_top_mistakes(comparisons)

    feedback = generate_feedback(
        reference_landmarks,
        live_landmarks,
        mistakes
    )

    return {

        "score": score,

        "worst_joints": [
            item["joint"]
            for item in mistakes
        ],

        "feedback": feedback

    }        
    

File: compare.js
// =========================================
// PoseSync - Pose Comparison Engine V2
// =========================================

// -----------------------------------------
// SETTINGS
// -----------------------------------------

const VISIBILITY_THRESHOLD = 0.5;

const MAX_DISTANCE = 0.02;

// Only compare important body joints
const IMPORTANT_JOINTS = [
    11, 12, // shoulders
    13, 14, // elbows
    15, 16, // wrists
    23, 24, // hips
    25, 26, // knees
    27, 28  // ankles
];

// Higher = more important
const JOINT_WEIGHTS = {

    11:2,
    12:2,

    13:3,
    14:3,

    15:4,
    16:4,

    23:2,
    24:2,

    25:3,
    26:3,

    27:4,
    28:4

};

const LANDMARK_NAMES = {

    11:"Left Shoulder",
    12:"Right Shoulder",

    13:"Left Elbow",
    14:"Right Elbow",

    15:"Left Wrist",
    16:"Right Wrist",

    23:"Left Hip",
    24:"Right Hip",

    25:"Left Knee",
    26:"Right Knee",

    27:"Left Ankle",
    28:"Right Ankle"

};

// -----------------------------------------
// Distance
// -----------------------------------------

function distance(a,b){

    return Math.sqrt(

        (a.x-b.x)**2 +

        (a.y-b.y)**2

    );

}

// -----------------------------------------
// Torso length
// -----------------------------------------

function torsoLength(landmarks){

    const leftShoulder = landmarks[11];

    const rightShoulder = landmarks[12];

    const leftHip = landmarks[23];

    const rightHip = landmarks[24];

    const shoulderCenter = {

        x:(leftShoulder.x+rightShoulder.x)/2,

        y:(leftShoulder.y+rightShoulder.y)/2

    };

    const hipCenter = {

        x:(leftHip.x+rightHip.x)/2,

        y:(leftHip.y+rightHip.y)/2

    };

    return distance(

        shoulderCenter,

        hipCenter

    );

}

// -----------------------------------------
// Normalize Pose
// -----------------------------------------

function normalizeLandmarks(landmarks){

    const leftHip = landmarks[23];

    const rightHip = landmarks[24];

    const centerX =

        (leftHip.x+rightHip.x)/2;

    const centerY =

        (leftHip.y+rightHip.y)/2;

    const scale = torsoLength(landmarks);

    return landmarks.map(point=>({

        x:(point.x-centerX)/scale,

        y:(point.y-centerY)/scale,

        visibility:point.visibility ?? 1

    }));

}

// -----------------------------------------
// Landmark validation
// -----------------------------------------

function validPose(landmarks){

    if(!landmarks) return false;

    if(landmarks.length!==33) return false;

    return true;

}
// =========================================
// ANGLE CALCULATIONS
// =========================================

// Returns the angle ABC in degrees
function calculateAngle(a, b, c) {

    const AB = {
        x: a.x - b.x,
        y: a.y - b.y
    };

    const CB = {
        x: c.x - b.x,
        y: c.y - b.y
    };

    const dot = (AB.x * CB.x) + (AB.y * CB.y);

    const magAB = Math.sqrt(AB.x * AB.x + AB.y * AB.y);
    const magCB = Math.sqrt(CB.x * CB.x + CB.y * CB.y);

    if (magAB === 0 || magCB === 0) return 0;

    let cosine = dot / (magAB * magCB);

    cosine = Math.max(-1, Math.min(1, cosine));

    return Math.acos(cosine) * 180 / Math.PI;

}

// -----------------------------------------
// Body angles we care about
// -----------------------------------------

function getPoseAngles(landmarks){

    return {

        leftArm:
            calculateAngle(
                landmarks[11],
                landmarks[13],
                landmarks[15]
            ),

        rightArm:
            calculateAngle(
                landmarks[12],
                landmarks[14],
                landmarks[16]
            ),

        leftLeg:
            calculateAngle(
                landmarks[23],
                landmarks[25],
                landmarks[27]
            ),

        rightLeg:
            calculateAngle(
                landmarks[24],
                landmarks[26],
                landmarks[28]
            ),

        leftTorso:
            calculateAngle(
                landmarks[11],
                landmarks[23],
                landmarks[25]
            ),

        rightTorso:
            calculateAngle(
                landmarks[12],
                landmarks[24],
                landmarks[26]
            )

    };

}

// -----------------------------------------
// Compare all angles
// -----------------------------------------

function compareAngles(reference, live){

    const refAngles = getPoseAngles(reference);

    const liveAngles = getPoseAngles(live);

    let totalDifference = 0;

    let count = 0;

    const mistakes = [];

    for(const joint in refAngles){

        const difference = Math.abs(

            refAngles[joint] -

            liveAngles[joint]

        );

        totalDifference += difference;

        count++;

        mistakes.push({

            joint,

            difference

        });

    }

    mistakes.sort(

        (a,b)=>b.difference-a.difference

    );

    return {

        averageDifference:

            totalDifference/count,

        mistakes

    };

}
// =========================================
// LANDMARK DISTANCE COMPARISON
// =========================================

function compareLandmarks(reference, live){

    const ref = normalizeLandmarks(reference);

    const current = normalizeLandmarks(live);

    let totalDistance = 0;

    let totalWeight = 0;

    const mistakes = [];

    for(const index of IMPORTANT_JOINTS){

        if(
            ref[index].visibility < VISIBILITY_THRESHOLD ||
            current[index].visibility < VISIBILITY_THRESHOLD
        ){
            continue;
        }

        const d = distance(

            ref[index],

            current[index]

        );

        const weight = JOINT_WEIGHTS[index];

        totalDistance += d * weight;

        totalWeight += weight;

        mistakes.push({

            index,

            joint: LANDMARK_NAMES[index],

            distance: d

        });

    }

    mistakes.sort(

        (a,b)=>b.distance-a.distance

    );

    return {

        averageDistance:

            totalDistance / totalWeight,

        mistakes

    };

}

// =========================================
// SCORE CALCULATION
// =========================================

function calculateScore(angleDiff, landmarkDiff){

    // ---------- ANGLES ----------

    let angleScore =

        100 -

        (angleDiff / 90) * 100;

    angleScore = Math.max(

        0,

        Math.min(100, angleScore)

    );

    // ---------- LANDMARKS ----------

    let landmarkScore =

        100 -

        (landmarkDiff / MAX_DISTANCE) * 100;

    landmarkScore = Math.max(

        0,

        Math.min(100, landmarkScore)

    );

    // ---------- COMBINED SCORE ----------

    let score =

        (0.85 * angleScore) +

        (0.15 * landmarkScore);

    // ---------- STRICTER SCORING ----------

    if(score > 90){

        score =

            90 +

            (score-90)*0.5;

    }

    else if(score > 70){

        score =

            70 +

            (score-70)*0.7;

    }

    else if(score > 50){

        score =

            50 +

            (score-50)*0.8;

    }

    return Number(

        score.toFixed(2)

    );

}
